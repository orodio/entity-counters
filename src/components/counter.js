import { html } from "@fp/html"
import { silence } from "@fp/silence"
import withCounter from "../sidecars/with-counter"
import counter from "../domains/counter"

export const Render = ({ counterId, count, inc = counter.inc, dec = counter.dec }) =>
  counterId == null
    ? html`
        <div>Loading...<//>
      `
    : html`
        <div>
          <div>
            <strong>${counterId}: <//>
            <span>${count}<//>
          <//>
          <div>
            <button onClick=${silence(_ => dec(counterId))}>dec<//>
            <button onClick=${silence(_ => inc(counterId))}>inc<//>
          <//>
        <//>
      `

export default withCounter(Render)
