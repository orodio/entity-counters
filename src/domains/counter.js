import { Entity } from "@fp/entity"
import { pick } from "@fp/pick"
import { getCount } from "./batch-fetch-counter"

const saveCountToApi = ({ id, count }) => Promise.resolve({ id, count })

export default Entity("counter", {
  withName: (id, label) => `${label}|${id}`,

  dependencies() {
    return {
      getCount,
      saveCount: saveCountToApi,
    }
  },

  broadcast: {
    count: state => pick(state, ["id", "count"]),
  },

  async init({ $$, Ok }, id) {
    const subs = new Set()
    const count = await $$.getCount(id)
    return Ok({ id, subs, count })
  },

  handleTell: {
    subscribe({ Ok, broadcast }, state, sub) {
      state.subs.add(sub)
      broadcast.count([sub], state)
      return Ok(state)
    },
    unsubscribe({ Ok }, state, sub) {
      state.subs.delete(sub)
      return Ok(state)
    },
    inc({ Ok, Continue, broadcast }, state, delta = 1) {
      state.count = state.count + delta
      broadcast.count(state.subs, state)
      return Ok(state, Continue("persist"))
    },
    dec({ Ok, Continue, broadcast }, state, delta = 1) {
      state.count = state.count - delta
      broadcast.count(state.subs, state)
      return Ok(state, Continue("persist"))
    },
  },

  handleContinue: {
    persist({ $$, Ok }, state) {
      $$.saveCount(pick(state, ["id", "count"]))
      return Ok(state)
    },
  },

  terminate(_ctx, _state, reason) {
    return reason
  },
})
