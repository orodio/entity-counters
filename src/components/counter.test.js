import { html } from "@fp/html"
import renderer from "react-test-renderer"
import counter from "../domains/counter"
import Counter, { Render } from "./counter"

const sleep = ms => new Promise(resolve => setTimeout(() => resolve(ms), ms))

test("Render", () => {
  const inc = jest.fn()
  const dec = jest.fn()

  const c1 = renderer.create(html`
    <${Counter} />
  `)
  expect(c1.toJSON()).toMatchSnapshot("c1 - no counterId - loading")

  const c2 = renderer.create(html`
    <${Counter} counterId="baz" />
  `)
  expect(c2.toJSON()).toMatchSnapshot("c2 - no count - loading")

  const c3 = renderer.create(html`
    <${Counter} counterId="baz" count=${9} inc=${inc} dec=${dec} />
  `)
  expect(c3.toJSON()).toMatchSnapshot("c3 - renders counter")
})

test("Counter", async () => {
  const ID = "baz"

  counter.init(ID, {
    debug: false,
    inject: {
      $$: {
        getCount: jest.fn(() => Promise.resolve(32)),
        saveCount: jest.fn(v => v),
      },
    },
  })

  let c1 = renderer.create(
    html`
      <${Counter} counterId=${ID} />
    `
  )

  expect(c1.toJSON()).toMatchSnapshot("c1 loading")
  await sleep(50)
  expect(c1.toJSON()).toMatchSnapshot("c1 after subscribe")

  counter.inc(ID)
  await sleep(50)
  expect(c1.toJSON()).toMatchSnapshot("c1 after inc")
})
