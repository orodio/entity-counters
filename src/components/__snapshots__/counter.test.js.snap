// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter: c1 after inc 1`] = `
<div>
  <div>
    <strong>
      baz
      : 
    </strong>
    <span>
      33
    </span>
  </div>
  <div>
    <button
      onClick={[Function]}
    >
      dec
    </button>
    <button
      onClick={[Function]}
    >
      inc
    </button>
  </div>
</div>
`;

exports[`Counter: c1 after subscribe 1`] = `
<div>
  <div>
    <strong>
      baz
      : 
    </strong>
    <span>
      32
    </span>
  </div>
  <div>
    <button
      onClick={[Function]}
    >
      dec
    </button>
    <button
      onClick={[Function]}
    >
      inc
    </button>
  </div>
</div>
`;

exports[`Counter: c1 loading 1`] = `
<div>
  <div>
    <strong>
      baz
      : 
    </strong>
    <span />
  </div>
  <div>
    <button
      onClick={[Function]}
    >
      dec
    </button>
    <button
      onClick={[Function]}
    >
      inc
    </button>
  </div>
</div>
`;

exports[`Render: c1 - no counterId - loading 1`] = `
<div>
  Loading...
</div>
`;

exports[`Render: c2 - no count - loading 1`] = `
<div>
  <div>
    <strong>
      baz
      : 
    </strong>
    <span />
  </div>
  <div>
    <button
      onClick={[Function]}
    >
      dec
    </button>
    <button
      onClick={[Function]}
    >
      inc
    </button>
  </div>
</div>
`;

exports[`Render: c3 - renders counter 1`] = `
<div>
  <div>
    <strong>
      baz
      : 
    </strong>
    <span>
      9
    </span>
  </div>
  <div>
    <button
      onClick={[Function]}
    >
      dec
    </button>
    <button
      onClick={[Function]}
    >
      inc
    </button>
  </div>
</div>
`;
